smoke-test/urls.test.js

```diff
@@ -0,0 +1,52 @@
+const {readFileSync} = require('fs');
+const testCases = [];
+
+// Load a suite of smoke tests from a simple text file. File is of the format:
+// [path], [comma separated text assertions]
+const text = readFileSync(require.resolve('./urls.txt'), 'utf8').trim();
+for (const line of text.split(/\r?\n/)) {
+  if (line.length === 0) continue; // allow for blank lines.
+  if (line.match(/^#/)) continue; // allow for comments.
+  const testCase = {assertions: []};
+  line.split(/,/).forEach((token, i) => {
+    if (i === 0) {
+      // The first token is the path that we wish to traverse the headless
+      // browser to:
+      testCase.path = token;
+    } else {
+      // The remaining tokens represent blocks of text in the rendered page
+      // we should check for:
+      testCase.assertions.push(token.trimLeft());
+    }
+  });
+  testCases.push(testCase);
+}
+
+describe("smoke test rendering of paths in urls.txt", () => {
+  for (const testCase of testCases) {
+    // eslint-disable-next-line no-loop-func
+    it(`it appropriately renders ${testCase.path}`, async () => {
+      const resp = await page.goto(`${BASE_URL}${testCase.path}`, { waitUntil: 'networkidle2' });
+      expect(resp.status()).toEqual(200);
+      // Create a Devtools session for fetching MIME HTML snapshot:
+      const client = await page.target().createCDPSession();
+      await client.send('Page.enable');
+      // Takes a MIME HTML snapshot of page, and makes sure it contains
+      // our smoke test text content:
+      const snapshotText = cleanSnapshotText(JSON.stringify(await client.send('Page.captureSnapshot')));
+      for (const assertion of testCase.assertions) {
+        expect(snapshotText).toEqual(expect.stringContaining(assertion));
+      }
+    });
+  }
+});
+
+// MIME HTML might have lines terminated with =\r\n, combine these lines
+// back together:
+function cleanSnapshotText(snapshotText) {
```

Comment:
This is nice. `snapshotText` also seems to be in [QP encoding](https://en.wikipedia.org/wiki/Quoted-printable) (not sure why) so it'd be nice to convert this to UTF-8 (I think). I couldn't find an out-of-the box node call to do this. At the lease we should replace `=3D` with `=` so we can do string contains checks such as `id="PhylogenyCard"` (although those quotations are escaped in `snapshotText` 🤷‍♂️ ). 

As a probably overly-complicated solution this could be parsed into a DOM tree, so then you could query things like (pseudocode) `hasDivWithId("PhylogenyCard")`. But I'm guessing this is getting outside the realm of smoke testing?!?

Reply:
As I understand it, the MHTML snapshot returns all the text content on the page (regardless of whether it's in a hidden div, frame, iframe, etc.) in MIME format. This is why it's QP Encoded (because it's a bunch of MIME content).

The nice thing, is you get a cheap snapshot of all of the visual text on the page.

> As a probably overly-complicated solution this could be parsed into a DOM tree

My concern with this, is the smoke tests start getting pretty complicated, and have to have semantic knowledge of the page structure, and as you change page structure these simple smoke tests need to change.

The integration tests in the `test/integration` folder also render the actual site using puppeteer, and are testing against specific elements in the page. It might be worth fleshing some of these tests out further, for specific examples like `hasDivWithId("PhylogenyCard")`, and if you do decide to have smoke tests as well, keep these assertions as simple as possible -- perhaps just checking for a single string in the whole page content that fingerprints the page.

Reply:
Thanks for humoring me 😄  I agree that it's best to keep smoke test nice and simple (especially since we can have them _right now_), and develop more complex logic in the integration tests.